### ISSUE TITLE
[Bug]: Firefox turns unresponsive after changing the camera permission while using it in the app.

### ISSUE THREAD

**Issue by amocirean** [mozilla-mobile/fenix#27228]
### Steps to reproduce

1. Open Firefox
2. Open the **QR Scanner**.
3. Navigate to the **Android's** app settings and tap on Firefox.
4. Change the Camera permission to **"Ask Every Time"**.
5. Return to Firefox.
6. Observe.

### Expected behaviour

The App should **ask again** for camera permission.

### Actual behaviour

The app turns **unresponsive**.

### Device name

Google Pixel 6

### Android version

Android 13

### Firefox release type

Firefox Beta

### Firefox version

106.0b3

### Additional information

Tested on both Pixel 6 (Android 13) and Pixel 4 (Android 12).

https://user-images.githubusercontent.com/103257761/193246698-e252cdfd-35b1-4d25-bf15-2c2ba8e47aa7.mp4

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21697)

**Comment by Mugurell** [comment]
This seems to be the same issue as on https://github.com/mozilla-mobile/fenix/issues/27177

**Comment by mavduevskiy** [comment]
It looks like it has been around since the introduction of the `ask every time` [option](https://support.google.com/android/answer/9431959?hl=en) in android11. I was able to reproduce the same behavior on the previous (105) build.

**Comment by mavduevskiy** [comment]
Upon some digging into the issue, it looks like after the introduction of [one-time-permissions](https://developer.android.com/training/permissions/requesting#one-time) handling permission edge cases has become hacky. one-time-permissions keeps `isPermissionGranted(Manifest.permission.CAMERA) == true` for a limited time. Restarting app (or sometimes the device) cases `isPermissionGranted(Manifest.permission.CAMERA)` to return false again. So the logic like this, which we currently have:
```
if (isPermissionGranted(CAMERA)) {
   qrFeature.scan()
} else {
  onCameraPermissionsNeeded()
}
``` 
 doesn't work as intended, it doesn't tell us when the rationale window should be displayed correctly. Starting from android11 denied permission requires to call permission window, not showing the rationale. 

Showing rationale is becomes even trickier because starting with android11 permission dialog goes into [denided-forever](https://developer.android.com/about/versions/11/privacy/permissions#dialog-visibility) mode and the system does not communicate that directly. The old `onRequestPermissionsResult` and the new `registerForActivityResult` just tell us if the permission has been granted or not, but doesn't tell if the permission window has been shown.

It looks like, apart from spying on the `GrantPermissionsActivity` directly, we could only rely on the method [shouldShowRequestPermissionRationale](https://developer.android.com/reference/androidx/core/app/ActivityCompat#shouldShowRequestPermissionRationale(android.app.Activity,%20java.lang.String)). It returns `true` after `don't allow` has been pressed. Two `don't allow` lead to the `denided-forever` state, so combination of `shouldShowRequestPermissionRationale == true` and `isGranted == false` should signalize the `denied-forever` state.

But given that we have to add flags to keep track of it and their state should be maintained properly (permissions could be reset from the system app settings widnow), I am not really excited about this approach.

**Comment by Mugurell** [comment]
Seems like the scenario for this issue happens because upon resuming from Settings the `QRFragment` from AC does not check again for the camera permission. 
So as an easy fix we can have the `QRFeature` check and ask again for the camera permission.

**Comment by Mugurell** [comment]
The main fix for this issue and for the ANR from https://github.com/mozilla-mobile/fenix/issues/27177#issuecomment-1263631497 is to happen in https://github.com/mozilla-mobile/android-components/issues/12882.

**Comment by Mugurell** [comment]
https://github.com/mozilla-mobile/android-components/pull/12883 was merged and helps with the app not becoming unresponsive.
https://github.com/mozilla-mobile/fenix/pull/27264 still waits for review for the app to check again the permissions after changing them in system settings, otherwise currently the QR scanner will show a black screen.

**Comment by SoftVision-LorandJanos** [comment]
Verified as fixed on the latest Nightly 109.0a1 (2022-11-22) build.
The prompt to allow firefox to take pictures and record video is displayed after the user changes the camera permission to "Ask Every Time" and returns to the app with the QR scanner on.
Device used: Oppo Find X5.
Closing the ticket as verified.

**Comment by amocirean** [comment]
Verified as fixed on the Firefox Beta 109.0b1 build.
